src/libs/actions/App.js

```diff
@@ -116,45 +110,49 @@ AppState.addEventListener('change', (nextAppState) => {
  * Fetches data needed for app initialization
  */
 function openApp() {
-    API.read('OpenApp', {policyIDList}, {
-        optimisticData: [{
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
-            key: ONYXKEYS.IS_LOADING_REPORT_DATA,
-            value: true,
-        }],
-        successData: [{
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
-            key: ONYXKEYS.IS_LOADING_REPORT_DATA,
-            value: false,
-        }],
-        failureData: [{
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
-            key: ONYXKEYS.IS_LOADING_REPORT_DATA,
-            value: false,
-        }],
+    policyIDListExcludingWorkspacesCreatedOfflineLoaded.then(() => {
+        API.read('OpenApp', {policyIDList: policyIDListExcludingWorkspacesCreatedOffline}, {
+            optimisticData: [{
+                onyxMethod: CONST.ONYX.METHOD.MERGE,
+                key: ONYXKEYS.IS_LOADING_REPORT_DATA,
+                value: true,
+            }],
+            successData: [{
+                onyxMethod: CONST.ONYX.METHOD.MERGE,
+                key: ONYXKEYS.IS_LOADING_REPORT_DATA,
+                value: false,
+            }],
+            failureData: [{
+                onyxMethod: CONST.ONYX.METHOD.MERGE,
+                key: ONYXKEYS.IS_LOADING_REPORT_DATA,
+                value: false,
+            }],
+        });
     });
 }
 
 /**
  * Refreshes data when the app reconnects
  */
 function reconnectApp() {
-    API.write('ReconnectApp', {policyIDListExcludingWorkspacesCreatedOffline}, {
-        optimisticData: [{
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
-            key: ONYXKEYS.IS_LOADING_REPORT_DATA,
-            value: true,
-        }],
-        successData: [{
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
-            key: ONYXKEYS.IS_LOADING_REPORT_DATA,
-            value: false,
-        }],
-        failureData: [{
-            onyxMethod: CONST.ONYX.METHOD.MERGE,
-            key: ONYXKEYS.IS_LOADING_REPORT_DATA,
-            value: false,
-        }],
+    policyIDListExcludingWorkspacesCreatedOfflineLoaded.then(() => {
+        API.write('ReconnectApp', {policyIDList: policyIDListExcludingWorkspacesCreatedOffline}, {
```

Comment:
Instead of waiting on a promise and then accessing the value separately, I think the promise should return the value itself. So ideally we would have something like the following
```javascript
getPolicyIdList()
    .then((policyIdList) =>  API.write('ReconnectApp', {policyIDList}, ...)
```

To do this I think you'd have to change the promise to `Onyx.connect` to get the value and `Onyx.disconnect` to clean up.

Reply:
Hmm, I agree with that it looks nicer, but this may not be so straight forward

- if the code stayed like it is and we return the value in the `resolve` and if `ReconnectApp` gets called a second time, it could see an outdated value because the subsequent resolves don't do anything.
- if we add a `getPolicyIdList()` that creates a promise each time, I don't see how you can reliably make sure that the promise will get resolved. Currently the promise only gets resolved when Onyx connect callback is called, but this won't happen if there is no change in the collection.

Reply:
If you want to push this path, could you help me with developing more your suggestion? I really don't see how this could be done that way at the moment... thanks!

Reply:
> Currently the promise only gets resolved when Onyx connect callback is called, but this won't happen if there is no change in the collection.

I had this concern as well. It's not straight forward to just get a value from Onyx I think the only way is to subscribe a callback with Onyx.connect. However, when you subscribe a new callback shouldn't the callback be called with the current value? If so, to get a value I think you should be able write a function like this:

```javascript
function getPolicyIDList () {
    return new Promise((resolve) => {
        const connectionID = Onyx.connect({ // temporary connection
            key: ONYXKEYS.COLLECTION.POLICY,
            waitForCollectionCallback: true,
            callback: (policies) => {
                const policyIDList = _.map(policies, policy => policy.id);
                Onyx.disconnect(connectionID); // cleanup connection
                resolve(policyIDList); // resolve with value
            },
        });
    });
}
```

It's a little messy with the temp connection (Ideally we could just `Onyx.get()` or something) but it should work the way you'd normally expect a promise to work. Do you mind trying if that works?

Reply:
Thanks @arosiclair for the complete example... sounds like it could work. I looked around our code and found the pattern you are proposing in a couple of places:

- https://github.com/Expensify/App/blob/main/src/libs/migrations/AddEncryptedAuthToken.js#L14-L22
- https://github.com/Expensify/App/blob/main/src/libs/E2E/actions/e2eLogin.js#L16-L50
- https://github.com/Expensify/App/blob/main/src/libs/migrations/RenamePriorityModeKey.js
- etc...

I have a concern about performance though, is it fast to read data from Onyx like that? If it is slow, I don't think it will be a problem if `getPolicyIDList()` is used only for cases like `OpenApp` or `ReconnectApp`, but if it is called from a render method, it could impact the app performance.

@roryabraham @marcaaron can I get your opinion here? 🙏

Reply:
Yeah I agree this probably should never be called from a render method. To prevent that we could just not export the function

Reply:
Thanks, I didn't know about the caches.

> maybe you can help me understand what we are trying to do and why it's important to know how fast the data will be read?

I just had a concern that adding such function could eventually end up in someone using it in rendering code hindering performance, but, thinking again about my concern, I think it wasn't on point because the `getPolicyIDList()` would really be async (return a promise), so I don't think the "performance" would a problem. The only "problem" in such case would be a new pattern for getting policies leaking into our components logic.

This last "problem" is something that can be assessed by leaving a comment and not exporting it.

Reply:
So.. considering the conversation https://github.com/Expensify/App/pull/12129/files#r1015744356, which made me realize that this really only needed for OpenApp, and not ReconnectApp, it sounds less useful to make a function `getPolicyIDList` if it is only going to be used in one place